handler/src/main/java/io/netty5/handler/ssl/EngineWrapper.java

```
@@ -0,0 +1,247 @@
+import io.netty5.buffer.api.internal.Statics;
+import io.netty5.util.internal.PlatformDependent;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import java.lang.ref.Reference;
+import java.nio.ByteBuffer;
+
+/**
+ * Used by {@link SslHandler} to {@linkplain SSLEngine#wrap(ByteBuffer, ByteBuffer) wrap} and
+ * {@linkplain SSLEngine#unwrap(ByteBuffer, ByteBuffer) unwrap} {@link Buffer} instances, which requires them to be
+ * projected into {@link ByteBuffer}s.
+ * <p>
+ * Instances of this class are single-threaded, and each {@link SslHandler} will have their own instance.
+ */
+class EngineWrapper implements ReadableComponentProcessor<RuntimeException>,
+                               WritableComponentProcessor<RuntimeException> {
+    /**
+     * Used for handshakes with engines that require off-heap buffers.
+     */
+    private static final ByteBuffer EMPTY_BUFFER_DIRECT = ByteBuffer.allocateDirect(0);
+    /**
+     * Used for handshakes with engines that require on-heap buffers.
+     */
+    private static final ByteBuffer EMPTY_BUFFER_HEAP = ByteBuffer.allocate(0);
+
+    private final SSLEngine engine;
+    private final boolean useDirectBuffer;
+
+    /**
+     * Used if {@link SSLEngine#wrap(ByteBuffer[], ByteBuffer)} and {@link SSLEngine#unwrap(ByteBuffer, ByteBuffer[])}
+     * should be called with a {@link ByteBuf} that is only backed by one {@link ByteBuffer} to reduce the object
+     * creation.
+     */
+    private final ByteBuffer[] singleEmptyBuffer;
+    private final ByteBuffer[] singleReadableBuffer;
+    private final ByteBuffer[] singleWritableBuffer;
+
+    private ByteBuffer[] inputs;
+    private ByteBuffer[] outputs;
+    private SSLEngineResult result;
+    private ByteBuffer cachedReadingBuffer;
+    private ByteBuffer cachedWritingBuffer;
+    private boolean writeBack;
+
+    EngineWrapper(SSLEngine engine, boolean useDirectBuffer) {
+        this.engine = engine;
```

should we add a `Objects.requireNonNull(...)` to fail fast ?